app/src/main/java/org/michiganhackers/photoassassin/LoginPages/RegistrationActivity.java

```diff
@@ -167,29 +172,84 @@ private void initializeUser(final String userId) {
 
         final Uri profilePicUri = getIntent().getParcelableExtra(PROFILE_PIC_URI);
         final String displayName = getIntent().getStringExtra(DISPLAY_NAME);
+        final String username = getIntent().getStringExtra(USERNAME);
+
+        Map<String, Object> displayNameMap = new HashMap<>();
+        displayNameMap.put("displayName", displayName);
+        displayNameMap.put("username", username);
+        FirebaseFunctions.getInstance()
+                .getHttpsCallable("addUser")
+                .call(displayNameMap)
+                .addOnCompleteListener(this, new OnCompleteListener<HttpsCallableResult>() {
+                    @Override
+                    public void onComplete(@NonNull Task<HttpsCallableResult> task) {
+                        if (task.isSuccessful()) {
+                            if (task.getResult() == null || task.getResult().getData() == null) {
+                                Log.e(TAG, "addUser cloud function didn't return result");
+                                return;
+                            }
+                            Map<String, Object> data = (Map<String, Object>) task.getResult().getData();
+                            String errorCode = (String) data.get("errorCode");
+                            if (errorCode == null) {
+                                Log.e(TAG, "addUser result data doesn't contain 'errorCode' field");
+                            } else if (errorCode.equals(Constants.ErrorCode.OK)) {
+                                Log.d(TAG, "addUser returned OK");
+                                User.getProfilePicRef(userId).putFile(profilePicUri)
+                                        .addOnFailureListener(new OnFailureListener() {
+                                            @Override
+                                            public void onFailure(@NonNull Exception e) {
+                                                Log.e(TAG, "Failed to add user profile pic to storage", e);
+
+                                            }
+                                        });
+                                Intent intent = new Intent(RegistrationActivity.this, MainActivity.class);
+                                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+                                startActivity(intent);
+                            } else if (errorCode.equals(Constants.ErrorCode.DUPLICATE_USERNAME)) {
+                                Log.w(TAG, "addUser returned DUPLICATE_USERNAME");
+                               handleDuplicateUsername();
```

Comment:
One suggestion is to handle duplicate usernames when the continue button is clicked in SetupProfileActivity. That way the user will know the username is taken without needing to input an email and password in the following registration activity screen.

Reply:
Okay, I can check for duplicate usernames in the setup profile activity in addition to the registration activity. The problem with just checking in the setup profile activity is that the user isn't actually created until the registration activity, and so the username could have been taken by someone else at that point. Thomas and I talked about a solution which almost always just requires checking in the setup profile activity by granting a lease for the username for a short amount of time [here](https://michiganhackers.slack.com/archives/CN9GYUF60/p1574298112000700), though we agreed that for v1 this lease solution isn't necessary.